compiler/src/iree/compiler/Codegen/Common/GPU/GPUCombineValueBarriers.cpp

```diff
@@ -0,0 +1,268 @@
+                        Operation *insertionPoint) {
+  SetVector<Operation *> toMove;
+  // First remove all operations that are already after the insertion point.
+  for (Operation *op : slice) {
+    if (op == insertionPoint) {
+      // A slice containing the insertion point itself cannot be moved
+      // behind the insertion point.
+      return failure();
+    }
+
+    if (op->isBeforeInBlock(insertionPoint)) {
+      toMove.insert(op);
+    }
+  }
+
+  // Sort operations to be moved topologically.
+  toMove = topologicalSort(toMove);
+
+  // It is always valid (w.r.t. dominance) to move topologically sorted
+  // operations in a forward slice which come before the insertion point, to
+  // after the insertion point. This is because:
+  //  - Since we are operating on a forward slice, consumers to every operation
+  //  in the slice are already in the slice, and will be moved after the
+  //  insertion point,
+  //  - Any producers will still remain before the operation, as we are only
+  //  moving the operation after.
+  for (Operation *sliceOp : llvm::reverse(toMove)) {
+    rewriter.moveOpAfter(sliceOp, insertionPoint);
+  }
+
+  return success();
+}
+
+/// Combine all value barriers into a single value barrier.
+static LogicalResult
+combineValueBarrierOps(RewriterBase &rewriter, Location loc,
+                       ArrayRef<IREE::GPU::ValueBarrierOp> valueBarriers) {
+  if (valueBarriers.size() <= 1) {
+    return success();
+  }
+  SmallVector<Value> barrierOperands;
+  for (auto barrierOp : valueBarriers) {
+    barrierOperands.append(barrierOp.getInputs().begin(),
+                           barrierOp.getInputs().end());
+  }
+  auto combinedBarrierOp =
+      rewriter.create<IREE::GPU::ValueBarrierOp>(loc, barrierOperands);
+
+  // Replace all uses of the previous barrier with new barrier.
+  int resultNumber = 0;
+  for (auto barrierOp : valueBarriers) {
+    int numResults = barrierOp.getNumResults();
+    rewriter.replaceOp(barrierOp, combinedBarrierOp->getResults().slice(
+                                      resultNumber, numResults));
+    resultNumber += numResults;
+  }
+  return success();
+}
+
+/// Given two barriers, barrierA and barrierB, combine them into a single
+/// barrier.
+static FailureOr<IREE::GPU::ValueBarrierOp>
+combineValueBarrierPair(RewriterBase &rewriter,
+                        IREE::GPU::ValueBarrierOp barrierA,
+                        IREE::GPU::ValueBarrierOp barrierB) {
+  // Both barriers need to have either tensor semantics or vector semantics.
+  if (barrierA.hasTensorSemantics() && !barrierB.hasTensorSemantics()) {
+    return failure();
+  }
+  if (!barrierA.hasTensorSemantics() && barrierB.hasTensorSemantics()) {
+    return failure();
+  }
+
+  // We assume barrierA is always before barrierB.
+  if (barrierB->isBeforeInBlock(barrierA)) {
+    std::swap(barrierA, barrierB);
+  }
+
+  // barrierA and barrierB are in the same block.
```

Comment:
Add an assert to check that?

Reply:
what is the result of `if (barrierB->isBeforeInBlock(barrierA))` when it _isn't_ in the same block?